src/libpmemobj/memops.c

```diff
@@ -440,6 +487,122 @@ operation_add_buffer(struct operation_context *ctx,
 			size - data_size, type);
 }
 
+/*
+ * operation_user_buffer_range_cmp -- compares addresses of
+ * user buffers
+ */
+int
+operation_user_buffer_range_cmp(const void *lhs, const void *rhs)
+{
+	const struct user_buffer_def *l = lhs;
+	const struct user_buffer_def *r = rhs;
+
+	if (l->addr > r->addr)
+		return 1;
+	else if (l->addr < r->addr)
+		return -1;
+
+	return 0;
+}
+
+/*
+ * operation_user_buffer_try_insert -- adds a user buffer range to the tree,
+ * if the buffer already exists in the tree function returns -1, otherwise
+ * it returns 0
+ */
+static int
+operation_user_buffer_try_insert(PMEMobjpool *pop, void *addr, size_t size)
+{
+	int ret = 0;
+
+	os_mutex_lock(&pop->ulog_user_buffers.lock);
+	struct user_buffer_def range;
+	range.addr = addr;
+	range.size = size;
+
+	if (ravl_emplace_copy(pop->ulog_user_buffers.map, &range) == -1 &&
+			errno == EEXIST)
+		ret = -1;
+
+	os_mutex_unlock(&pop->ulog_user_buffers.lock);
+	return ret;
+}
+
+/*
+ * operation_add_user_buffer -- add user buffer to the ulog
+ */
+int
+operation_add_user_buffer(struct operation_context *ctx,
+		void *addr, size_t size)
+{
+	uint64_t extend_offset = OBJ_PTR_TO_OFF(ctx->p_ops->base, addr);
```

Comment:
`Address of the buffer has to be aligned up, and the size has to be aligned down, taking into account the number of bytes the address was incremented by. The remaining size, has to be large enough to contain the header and at least one ulog entry.`